source/routing/heat_map.cpp

```diff
@@ -120,60 +81,152 @@ static void print_body(std::stringstream& ss,
+                    auto center = type::GeographicalCoord(heat_map.body[lon_rank].first.min_coord + width_step/2,
+                                                          heat_map.header[lat_rank].min_coord + height_step / 2);
+                    if(box.contains(center)) {
+                        auto proj = center.approx_project(source, target, coslat);
+                        if (!dist_pixel[lon_rank][lat_rank].distance ||
+                             (proj.second < min_dist && proj.second < *dist_pixel[lon_rank][lat_rank].distance)) {
+                            dist_pixel[lon_rank][lat_rank].distance = proj.second;
+                            dist_pixel[lon_rank][lat_rank].source = it->element;
+                            dist_pixel[lon_rank][lat_rank].target = v;
+                        }
+                    }
+                }
             }
-            auto start_coord = type::GeographicalCoord(lon.min_coord + width_step / 2, header[j].min_coord + height_step / 2);
-            auto projected_edge = georef::ProjectionData(start_coord, worker, offset, worker.pl, min_dist);
-            auto dur = set_duration(projected_edge, start_coord, distances, speed, max_duration);
-            local_duration.push_back(dur);
         }
-        auto local_body = std::make_pair(lon, local_duration);
-        body.push_back(std::move(local_body));
     }
-    return HeatMap(header, body);
+    return dist_pixel;
+}
+
+HeatMap fill_heat_map(BoundBox box,
+                      const double height_step,
+                      const double width_step,
+                      const georef::GeoRef& worker,
+                      const double min_dist,
+                      const double max_duration,
+                      const double speed,
+                      const std::vector<navitia::time_duration>& distances,
+                      const size_t step) {
+    auto heat_map = HeatMap(step, box, height_step, width_step);
+    auto projection = find_projection(box, height_step, width_step, worker, min_dist, heat_map, step);
+    for (size_t i = 0; i < step; i++){
+        for (size_t j = 0; j < step; j++){
+            auto& duration = heat_map.body[i].second[j];
+            if (projection[i][j].distance) {
+                auto center = type::GeographicalCoord(heat_map.body[i].first.min_coord + width_step/2,
+                                                      heat_map.header[j].min_coord + height_step / 2);
+                const auto source = worker.graph[projection[i][j].source].coord;
+                const auto target = worker.graph[projection[i][j].target].coord;
+                const auto coslat = center.lat() * type::GeographicalCoord::N_DEG_TO_RAD;
+                const auto duration_to_source = distances[projection[i][j].source] +
+                        navitia::milliseconds(sqrt(center.approx_sqr_distance(source, coslat)) / speed * 1e3);
```

Comment:
shouldn't `*projection[i][j].distance` be used instead of recomputing something?